sql/lease_test.go

```diff
@@ -0,0 +1,327 @@
+}
+
+func TestLeaseManagerReacquire(testingT *testing.T) {
+	defer leaktest.AfterTest(testingT)
+	t := newLeaseTest(testingT)
+	defer t.cleanup()
+
+	const descID = keys.LeaseTableID
+
+	// Acquire 2 leases from the same node. They should point to the same lease
+	// structure.
+	l1 := t.mustAcquire(1, descID, 0)
+	l2 := t.mustAcquire(1, descID, 0)
+	if l1 != l2 {
+		t.Fatalf("expected same lease, but found %p != %p", l1, l2)
+	}
+	if l1.Refcount() != 2 {
+		t.Fatalf("expected refcount of 2, but found %d", l1.Refcount())
+	}
+	t.expectLeases(descID, "/1/1")
+
+	// Push the clock up near the lease expiration.
+	t.server.Clock().Update(roachpb.Timestamp{
+		WallTime: l1.Expiration().Add(-20 * time.Second).UnixNano(),
+	})
+
+	// Another lease acquisition from the same node will result in a new lease.
+	l3 := t.mustAcquire(1, descID, 0)
+	if l1 == l3 {
+		t.Fatalf("expected different leases, but found %p", l1)
+	}
+	if l3.Refcount() != 1 {
+		t.Fatalf("expected refcount of 1, but found %d", l3.Refcount())
+	}
+	t.expectLeases(descID, "/1/1 /1/1")
+
+	t.mustRelease(1, l1)
+	t.mustRelease(1, l2)
+	t.mustRelease(1, l3)
+}
+
+func TestLeaseManagerPublishVersionChanged(testingT *testing.T) {
+	// This test takes ~10 seconds because of the unresolved write intents in the
+	// system DB span.
+	testingT.Skipf("TODO(tschottdorf): #2871")
+
+	defer leaktest.AfterTest(testingT)
+	t := newLeaseTest(testingT)
+	defer t.cleanup()
+
+	const descID = keys.LeaseTableID
+
+	// Start two goroutines that are concurrently trying to publish a new version
+	// of the descriptor. The first goroutine progresses to the update function
+	// and then signals the second goroutine to start which is allowed to proceed
+	// through completion. The first goroutine is then signalled and when it
+	// attempts to publish the new version it will encounter an update error and
+	// retry the transaction. Upon retry it will see that the descriptor version
+	// has changed and have to proceed to its outer retry loop and wait for the
+	// number of leases on the previous version to drop to 0.
+
+	n1 := t.node(1)
+	n2 := t.node(2)
+
+	n1update := make(chan struct{})
+	n2start := make(chan struct{})
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	go func(n1update, n2start chan struct{}) {
+		err := n1.Publish(descID, func(*csql.TableDescriptor) error {
+			if n2start != nil {
+				// Signal node 2 to start.
+				close(n2start)
+				n2start = nil
+			}
+			// Wait for node 2 signal indicating that node 2 finished publication of
+			// a new version.
+			<-n1update
+			return nil
+		})
+		if err != nil {
+			panic(err)
+		}
+		wg.Done()
+	}(n1update, n2start)
+
+	go func(n1update, n2start chan struct{}) {
+		// Wait for node 1 signal indicating that node 1 is in its update()
+		// function.
+		<-n2start
+		err := n2.Publish(descID, func(*csql.TableDescriptor) error {
+			return nil
+		})
+		if err != nil {
+			panic(err)
+		}
+		close(n1update)
+		wg.Done()
+	}(n1update, n2start)
+
+	wg.Wait()
+}
```

Comment:
anything being asserted in this test other than that nothing errors out? Or weren't you motivated to finish it with the system span issue?